firedrake/ensemble/ensemble_function.py

```diff
@@ -0,0 +1,292 @@
+from firedrake.petsc import PETSc
+from firedrake.ensemble.ensemble_functionspace import (
+    EnsembleFunctionSpace, EnsembleDualSpace)
+from firedrake.adjoint_utils import EnsembleFunctionMixin
+from firedrake.function import Function
+from firedrake.norms import norm
+from pyop2 import MixedDat
+
+from functools import cached_property
+from contextlib import contextmanager
+
+__all__ = ("EnsembleFunction", "EnsembleCofunction")
+
+
+class EnsembleFunctionBase(EnsembleFunctionMixin):
+    """
+    A mixed (co)function defined on a :class:`firedrake.Ensemble`.
+    The subcomponents are distributed over the ensemble members, and
+    are specified locally in a :class:`firedrake.EnsembleFunctionSpace`.
```

Comment:
Not sure what you mean by "are specified locally"

Reply:
Anywhere I say "locally" I mean on each `Ensemble.comm`. I should either define this somewhere or be more explicit in each place I currently use "locally".